ocaml/xapi/repository.ml

```diff
@@ -1062,42 +1062,84 @@ let assert_valid_guidances = function
   | [EvacuateHost]
   | [RestartToolstack]
   | [RestartDeviceModel] -> ()
-  | l when eq l set1 -> ()
-  | l when eq l set2 -> ()
-  | l when eq l set3 -> ()
-  | l when eq l set4 -> ()
+  | l when eq l set1 ->
+    (* EvacuateHost and RestartToolstack *)
+    ()
+  | l when eq l set2 ->
+    (* RestartDeviceModel and RestartToolstack *)
+    ()
+  | l when eq l set3 ->
+    (* RestartDeviceModel and EvacuateHost *)
+    ()
+  | l when eq l set4 ->
+    (* EvacuateHost, RestartToolstack and RestartDeviceModel *)
+    ()
   | l ->
     let msg =
       Printf.sprintf "Found wrong guidance(s) before applying updates: %s"
         (String.concat ";" (List.map string_of_guidance l))
     in
     raise Api_errors.(Server_error (internal_error, [msg]))
 
+let restart_device_models ~__context host =
+  (* Restart device models of all running HVM VMs on the host by doing
+   * local migrations. *)
+  Db.Host.get_resident_VMs ~__context ~self:host
+  |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+  |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
+  |> List.filter_map (fun (ref, _) ->
+      match Db.VM.get_power_state ~__context ~self:ref,
+            Helpers.has_qemu_currently ~__context ~self:ref  with
+      | `Running, true ->
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            Client.Client.VM.pool_migrate rpc session_id ref host [("live", "true")]);
+        None
+      | `Paused, true ->
+        error "VM 'ref=%s' is paused, can't restart its device models" (Ref.string_of ref);
+        Some ref
+      | _ ->
+        (* No device models are running for this VM *)
+        None)
+  |> function
+  | [] -> ()
+  | _ :: _ ->
```

Comment:
You could use `Option.iter` instead.

Reply:
Not quite understand this :(

`val iter : ('a -> unit) -> 'a option -> unit
iter f o is f v if o is Some v and () otherwise.`

Here we have a list of `ref option`, and we aim to raise exception if the list is not empty.

Reply:
Yes, I misread this.